plugins/usbdmx/Siudi.cpp

```diff
@@ -0,0 +1,155 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Siudi.cpp
+ * The synchronous SIUDI widgets.
+ * Copyright (C) 2023 Alexander Simon
+ */
+
+#include "plugins/usbdmx/Siudi.h"
+
+#include <string.h>
+#include <unistd.h>
+
+#include "libs/usb/LibUsbAdaptor.h"
+#include "ola/Constants.h"
+#include "ola/Logging.h"
+#include "plugins/usbdmx/AsyncUsbSender.h"
+#include "plugins/usbdmx/ThreadedUsbSender.h"
+
+namespace ola {
+namespace plugin {
+namespace usbdmx {
+
+using ola::usb::LibUsbAdaptor;
+
+namespace {
+
+static const uint8_t ENDPOINT = 2;
+static const unsigned int TIMEOUT = 50; // 50ms is ok
+enum {SIUDI_PACKET_SIZE = 512};
+
+}  // namespace
+
+// SiudiThreadedSender
+// -----------------------------------------------------------------------------
+
+/*
+ * Sends messages to a SIUDI device in a separate thread.
+ */
+class SiudiThreadedSender: public ThreadedUsbSender {
+ public:
+  SiudiThreadedSender(LibUsbAdaptor *adaptor,
+                        libusb_device *usb_device,
+                        libusb_device_handle *handle);
+
+  bool Start();
+
+private:
+  LibUsbAdaptor* const m_adaptor;
+  uint8_t m_packet[SIUDI_PACKET_SIZE];
+  libusb_device_handle* const m_usb_handle;
+
+  bool TransmitBuffer(libusb_device_handle *handle,
+                      const DmxBuffer &buffer);
+};
+
+SiudiThreadedSender::SiudiThreadedSender(
+    LibUsbAdaptor *adaptor,
+    libusb_device *usb_device,
+    libusb_device_handle *usb_handle)
+    : ThreadedUsbSender(usb_device, usb_handle),
+      m_adaptor(adaptor), m_usb_handle(usb_handle) {
+  memset(m_packet, 0x00, SIUDI_PACKET_SIZE);
+}
+
+bool SiudiThreadedSender::Start() {
+  if (!ThreadedUsbSender::Start()) {
+    return false;
+  }
+
+  // Read device info. This call takes about 270 ms.
+  // Discard the buffer as the format is currently unknown.
+  uint8_t buf[64];
+  int ret = libusb_control_transfer(m_usb_handle,
+                                LIBUSB_REQUEST_TYPE_VENDOR | LIBUSB_ENDPOINT_IN,
+                                0x3f, 0x00c4, 1, buf, 64, 500);
+  if (ret != 64) {
+    OLA_WARN << "Failed to read SIUDI information: "
+            << (ret < 0 ? LibUsbAdaptor::ErrorCodeToString(ret) : "Short read");
+    return false;
+  }
+
+  // Unstall the endpoint. The original software seems to call this regularly.
+  ret = libusb_clear_halt(m_usb_handle, ENDPOINT);
+  if (ret != 0) {
+    OLA_WARN << "Failed to reset SIUDI endpoint: "
+            << (ret < 0 ? LibUsbAdaptor::ErrorCodeToString(ret) : "Unknown");
+    return false;
+  }
+  usleep(10000);
+
+  return true;
+}
+
+bool SiudiThreadedSender::TransmitBuffer(libusb_device_handle *handle,
+                                           const DmxBuffer &buffer) {
+  for (unsigned int i = 0; i < buffer.Size(); i++) {
+    m_packet[i] = buffer.Get(i);
+  }
+  int transferred;
+  int r = m_adaptor->BulkTransfer(
+    handle, ENDPOINT, (unsigned char*) m_packet,
```

Comment:
But given you're just writing it untouched, you could probably just get away with GetRaw here:
http://docs.openlighting.org/ola/doc/latest/classola_1_1_dmx_buffer.html#a50eff7fa1c6a8d21ffa5ccb14a781de9

Reply:
Due to the fact that I always have to send 512 bytes, I could only use `GetRaw` when `DmxBuffer.Size()` was 512. If it differs, I used `DmxBuffer.GetRange()` to copy 512 or less bytes. If it is less than 512, the remaining bytes are `memset` to 0x00.

See 96e10f70e7d02ce70151973234a2cd4a310aeebf and a fixup in 42b23c745c3ef089f3543c4a098e51c9cd791af1.